cuqi/experimental/mcmc/_rto.py

```diff
@@ -0,0 +1,273 @@
+
+        # Check Linear model and Gaussian likelihood(s)
+        if isinstance(self.target, cuqi.distribution.Posterior):
+            if not isinstance(self.model, cuqi.model.LinearModel):
+                raise TypeError("Model needs to be linear")
+
+            if not hasattr(self.likelihood.distribution, "sqrtprec"):
+                raise TypeError("Distribution in Likelihood must contain a sqrtprec attribute")
+            
+        elif isinstance(self.target, cuqi.distribution.MultipleLikelihoodPosterior): # Elif used for further alternatives, e.g., stacked posterior
+            for likelihood in self.likelihoods:
+                if not isinstance(likelihood.model, cuqi.model.LinearModel):
+                    raise TypeError("Model needs to be linear")
+
+                if not hasattr(likelihood.distribution, "sqrtprec"):
+                    raise TypeError("Distribution in Likelihood must contain a sqrtprec attribute")
+        
+        # Check Gaussian prior
+        if not hasattr(self.prior, "sqrtprec"):
+            raise TypeError("prior must contain a sqrtprec attribute")
+
+        if not hasattr(self.prior, "sqrtprecTimesMean"):
+            raise TypeError("Prior must contain a sqrtprecTimesMean attribute")
+
+    def get_state(self): #TODO: LinearRTO only need initial_point for reproducibility?
+        return {'sampler_type': 'LinearRTO'}
+
+    def set_state(self, state): #TODO: LinearRTO only need initial_point for reproducibility?
+        pass
+
+class RegularizedLinearRTONew(LinearRTONew):
+    """
+    Regularized Linear RTO (Randomize-Then-Optimize) sampler.
+
+    Samples posterior related to the inverse problem with Gaussian likelihood and implicit Gaussian prior, and where the forward model is Linear.
+
+    Parameters
+    ------------
+    target : `cuqi.distribution.Posterior`
+        See `cuqi.sampler.LinearRTO`
+
+    initial_point : `np.ndarray` 
+        Initial point for the sampler. *Optional*.
+
+    maxit : int
+        Maximum number of iterations of the inner FISTA solver. *Optional*.
+        
+    stepsize : string or float
+        If stepsize is a string and equals either "automatic", then the stepsize is automatically estimated based on the spectral norm.
+        If stepsize is a float, then this stepsize is used.
+
+    abstol : float
+        Absolute tolerance of the inner FISTA solver. *Optional*.
+
+    callback : callable, *Optional*
+        If set this function will be called after every sample.
+        The signature of the callback function is `callback(sample, sample_index)`,
+        where `sample` is the current sample and `sample_index` is the index of the sample.
+        An example is shown in demos/demo31_callback.py.
+        
+    """
+    def __init__(self, target, maxit=100, stepsize = "automatic", abstol=1e-10, adaptive = True, **kwargs):
```

Comment:
- Added description on `adaptive` in docstring.
- As to `initial_point`, I think it is taken care of as part of `kwargs`, then we don't need to explicitly list it as an argument of the constructor? I'm confident here.

Reply:
Right, I see it is handled through the super call by the lines of LinearRTONew:
```
        if 'initial_point' not in kwargs:
            self.initial_point = np.zeros(self.dim)
```
Two questions then. 

Can this default value not be specified as as part of the `__init__` line? It would seem neater that `initial_point` is seen explicitly in the list of inputs, such that the list of inputs matches exactly the inputs stated in the docstring. Is it because `self.dim` needs to be set before we can create the zero vector of the correct size, and setting `self.dim` is handled by the `super` call from `LinearRTO`?

The `__init__` inputs to `LinearRTONew` and `RegularizedLinearRTONew` are slightly different
```
def __init__(self, target, maxit=10, tol=1e-6, shift=0, **kwargs)
```
vs
```
    def __init__(self, target, maxit=100, stepsize = "automatic", abstol=1e-10, adaptive = True, **kwargs):
```
Is that because, in the first one, `tol` and `shift` are just used for the CGLS, while for the second `stepsize`, `abstol` and `adaptive` are just used for FISTA?

Reply:
Done as suggested after group discussion